File: online_judge/AtCoder/ABC154/F/f2.cpp
#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

#define rep(i, n) for (int i = 0; i < (n); i++)

template <int64_t mod>
struct modint {
  int64_t a;
  modint(int64_t a = 0) {
    a %= mod;
    if (a < 0) a += mod;
    this->a = a;
  }
  modint operator+(modint rhs) { return modint(a + rhs.a); }
  modint operator-(modint rhs) { return modint(a - rhs.a); }
  modint operator*(modint rhs) { return modint(a * rhs.a); }
  modint pow(modint base, int64_t exp) {
    auto b = base;
    auto res = modint(1);
    while (exp > 0) {
      if (exp & 1) { res = res * b; }
      b = b * b;
      exp = exp / 2;
    }
    return res;
  }
  modint inv() { return pow(*this, mod - 2); }
  modint operator/(modint rhs) { return (*this) * rhs.inv(); }
};

using mint = modint<1000000000 + 7>;
mint solve(int r, int c) {
  auto n = r + c + 2;
  vector<mint> fac(n), inv(n);
  fac[0] = 1;
  for (int i = 1; i < n; i++) {
    // fac[i] = mint(i) * fac[i - 1];
    fac[i] = fac[i - 1] * i;
  }
  inv[n - 1] = fac[n - 1].inv();
  for (int i = n - 2; i >= 0; i--) {
    inv[i] = inv[i + 1] * (i + 1);
  }
  auto binom = [&](int a, int b) -> mint {
    if (a < 0 or b < 0) return 0;
    if (a < b) return 0;
    return fac[a] * inv[b] * inv[a - b];
  };
  auto countPath = [&](int a, int b) { return binom(a + b, b); };
  auto res = mint(0);
  for (int i = 0; i <= r; i++) {
    res = res + countPath(i + 1, c);
  }
  return res;
}

int main() {

  int r1, c1, r2, c2;
  cin >> r1 >> c1 >> r2 >> c2;

  auto ans = solve(r2, c2);
  ans = ans - solve(r1 - 1, c2);
  ans = ans - solve(r2, c1 - 1);
  ans = ans + solve(r1 - 1, c1 - 1);
  cout << ans.a << endl;

  return 0;
}

File: online_judge/AtCoder/ABC160/F/f2.nim
import strutils, sequtils, queues

const mo: int64 = 1000000000 + 7
type mint = distinct int64
proc toMint[Int](a: Int): mint =
  var res = a.int64 mod mo
  if res < 0:
    res = res + mo
  return res.mint
proc `+`(a, b: mint): mint = (a.int64 + b.int64).toMint
proc `-`(a, b: mint): mint = (a.int64 - b.int64).toMint
proc `*`(a, b: mint): mint = (a.int64 * b.int64).toMint
proc `*=`[Int](a: var mint, b: Int) = a = a * b.toMint
proc pow(a: mint, x: int64): mint =
  assert(x >= 0)
  var
    res = 1.toMint
    base = a
    exp = x
  while exp > 0:
    if (exp and 1) == 1:
      res = res * base
    base = base * base
    exp = exp div 2
  return res
proc inv(a: mint): mint = pow(a, mo - 2)
proc `div`(a, b: mint): mint = a * inv(b)
proc `$`(a: mint): string {.borrow.}

template yieldCombination(n: int = 20) =
  var
    fac {.inject.} = newSeq[mint](n)
    inv {.inject.} = newSeq[mint](n)
  fac[0] = 1.toMint
  for i in 1..<n:
    fac[i] = i.toMint * fac[i - 1]
  inv[n - 1] = 1.toMint div fac[n - 1]
  for i in countdown(n - 2, 0):
    inv[i] = inv[i + 1] * (i + 1).toMint
  let binom {.inject.} =
    proc(a, b: int): mint =
      if a < 0 or b < 0: return 0.toMint
      if a < b: return 0.toMint
      return fac[a] * inv[b] * inv[a - b]

let read = iterator: string {.closure.} =
  for s in stdin.readAll.split:
    yield s

proc main() =
  let n = read().parseInt
  var g = newSeqWith(n, newSeq[int]())
  for i in 1..<n:
    let a, b = read().parseInt
    g[a - 1].add(b - 1)
    g[b - 1].add(a - 1)

  yieldCombination(n + 1)
  # 頂点 1 を根とする
  # size[i]: 部分木 i に含まれる頂点数
  # dp[i]: 部分木 i 以下で考えたときの書き方の数
  #   = size[i] 個の頂点を親が左・子が右になる並べ方の数
  #     (左から 1, 2, ... と書き込む)
  #   i を一番左にする
  #   i のそれぞれの子 j について
  #     size[j] 個ずつ選んで合計 size[i] - 1 個選ぶ: 多項係数
  #   選んだ size[j] 個の並べ方は dp[j] 通り
  var
    size = newSeq[int](n)
    dp = newSeq[mint](n)
  proc dfs(i, p: int) =
    size[i] = 1
    dp[i] = 1.toMint
    for j in g[i]:
      if j != p:
        dfs(j, i)
        size[i] += size[j]
        dp[i] *= dp[j] div fac[size[j]]
    dp[i] *= fac[size[i] - 1]
  dfs(0, -1)
  # echo dp
  var ans = newSeq[mint](n)
  var q = initQueue[(int, int, int, mint)]()
  q.enqueue((0, -1, 0, 1.toMint))
  while q.len > 0:
    let (i, p, a, b) = q.dequeue
    var tot = 1.toMint
    for j in g[i]:
      tot *= (if j == p: b div fac[a] else: dp[j] div fac[size[j]])
    ans[i] = tot * fac[n - 1]
    for j in g[i]:
      if j != p:
        let na = a + size[i] - size[j]
        q.enqueue((j, i, na, tot div (dp[j] div fac[size[j]]) * fac[na - 1]))
  echo ans.mapIt($it).join("\n")
main()
